sapicore/utils/io.py

```diff
@@ -0,0 +1,270 @@
+                with hdf.File(os.path.join(self.log_dir, self.component.identifier + ".h5"), "a") as hf:
+                    for attr in self.attributes:
+                        # increase data chunk size by number of rows in the cache tensor.
+                        self.chunk_size[attr] += self.cache[attr].shape[0]
+                        hf[attr].resize(self.chunk_size[attr], axis=0)
+
+                        # write current cache tensor to its respective HDF5 data structure at the correct indices.
+                        start_index = hf[attr].shape[0] - self.cache[attr].shape[0]
+                        end_index = hf[attr].shape[0]
+
+                        # detach().numpy() necessary if the tensor requires grad.
+                        hf[attr][start_index:end_index, :] = self.cache[attr].cpu().detach().numpy()
+
+                    # clear cache.
+                    self.cache = {}
+
+        return fn
+
+    def forward(self, data: Tensor) -> None:
+        _ = self.component(data)
+        return
+
+
+def log_settings(
+    level: int = logging.INFO,
+    stream: _io.TextIOWrapper = sys.stdout,
+    file: str = None,
+    formatting: str = "%(asctime)s [%(levelname)s] %(message)s",
+):
+    """Temporary basic logger configuration.
+
+    Parameters
+    ----------
+    level: int
+        Logger level, `INFO` by default.
+
+    stream:
+        Destination stream, `stdout` by default.
+
+    file:
+        Destination file, `None` by default.
+
+    formatting: str
+        Message formatting.
+
+    """
+    logging.basicConfig(
+        level=level,
+        format=formatting,
+        handlers=[logging.StreamHandler(stream)] + ([logging.FileHandler(file)] if file else []),
+    )
+
+
+def load_apply_config(configuration: str, apply_to: object = None) -> dict:
+    # parse configuration from YAML if given and use resulting dictionary to initialize attributes.
+    content = load_config(None, configuration) if os.path.exists(configuration) else None
+
+    if not content:
+        raise FileNotFoundError(f"Could not find a configuration YAML at {os.path.realpath(configuration)}")
+
+    else:
+        # apply the configuration to this pipeline object.
+        configuration = content
+
+        if apply_to:
+            apply_config(apply_to, configuration)
+
+    return configuration
+
+
+def ensure_dir(path: str = None) -> str:
+    """Ensures that a path is created if parts thereof do not exist yet and returns it.
+
+    Parameters
+    ----------
+    path: str
+        File path to be verified or created.
+
+    Returns
+    -------
+    path: str
+        The same file path, so it can be used in the calling context (e.g., if :func:`ensure_dir`
+        wraps an assignment).
+
+    """
+    os.makedirs(path, exist_ok=True)
+    return path
+
+
+def flatten(lst: list[list]) -> list:
+    """Flattens a list."""
+    return [item for sublist in lst for item in sublist]
```

Comment:
When I was pointing to `itertools.chain` I meant the "traditional" functools way of doing this is with `list(chain(*lst))`.

Reply:
Changed